src/provider/mod.rs

```diff
@@ -0,0 +1,181 @@
+        };
+        let missing_crds = provider.check_crds().await;
+        if missing_crds.is_empty() {
+            debug!("All required CRDS present!");
+            return Ok(provider);
+        } else {
+            debug!("Missing required CDRS");
+            return Err(CrdMissing { missing_crds });
+        }
+    }
+
+    fn get_package(&self, pod: &Pod) -> Result<Package, StackableError> {
+        let containers = pod.containers();
+        if containers.len().ne(&1) {
+            let e = PodValidationError {
+                msg: String::from("Size of containers list in PodSpec has to be exactly 1"),
+            };
+            return Err(e);
+        } else {
+            // List has exactly one value, try to parse this
+            if let Ok(Some(reference)) = containers[0].image() {
+                return Package::try_from(reference);
+            } else {
+                let e = PodValidationError {
+                    msg: String::from("Unable to get package reference from pod"),
+                };
+                return Err(e);
+            }
+        }
+    }
+
+    async fn check_crds(&self) -> Vec<String> {
+        let mut missing_crds = vec![];
+        let crds: Api<CustomResourceDefinition> = Api::all(self.client.clone());
+
+        // Check all CRDS
+        for crd in CRDS.into_iter() {
+            debug!("Checking if CRD \"{}\" is registered", crd);
+            match crds.get(crd).await {
+                Err(e) => {
+                    error!("Missing required CRD: \"{}\"", crd);
+                    missing_crds.push(String::from(*crd))
+                }
+                _ => debug!("Found registered crd: {}", crd),
+            }
+        }
+        missing_crds
+    }
+}
+
+// No cleanup state needed, we clean up when dropping PodState.
+#[async_trait::async_trait]
+impl kubelet::state::AsyncDrop for PodState {
+    async fn async_drop(self) {}
+}
+
+#[async_trait::async_trait]
+impl Provider for StackableProvider {
+    type PodState = PodState;
+    type InitialState = Downloading;
+    type TerminatedState = Terminated;
+
+    const ARCH: &'static str = "stackable-linux";
+
+    async fn node(&self, builder: &mut Builder) -> anyhow::Result<()> {
+        builder.set_architecture(Self::ARCH);
+        builder.add_taint("NoSchedule", "kubernetes.io/arch", Self::ARCH);
+        builder.add_taint("NoExecute", "kubernetes.io/arch", Self::ARCH);
+        Ok(())
+    }
+
+    async fn initialize_pod_state(&self, pod: &Pod) -> anyhow::Result<Self::PodState> {
+        let service_name = pod.name();
+        let parcel_directory = self.parcel_directory.clone();
+        let download_directory = parcel_directory.join("_download");
+        let config_directory = self.config_directory.clone();
+        let log_directory = self.log_directory.clone();
+
+        let package = self.get_package(pod)?;
+        if !(&download_directory.is_dir()) {
+            fs::create_dir_all(&download_directory)?;
```

Comment:
It is called directly from the pod informer[1] in the kubelet, which would log an error if this fails, but otherwise continue.
I think we don't want to change this as otherwise one faulty pod might impact the entire agent. We do need to monitor this of course, this can happen via the pod/container state that is reported, or via metrics that we write out of the agent. We should look into adding this at some point.


[1] https://github.com/deislabs/krustlet/blob/master/crates/kubelet/src/kubelet.rs#L196